src/NUnitEngine/nunit.engine/Runners/WorkItemTracker.cs

```diff
@@ -0,0 +1,113 @@
+    /// in order to capture events. It allows waiting for all items to complete.
+    /// Once the test has been cancelled, it provide notifications to the runner
+    /// so the information may be displayed.
+    /// </summary>
+    internal class WorkItemTracker : ITestEventListener
+    {
+        private List<XmlNode> _itemsInProcess = new List<XmlNode>();
+        private ManualResetEvent _allItemsComplete = new ManualResetEvent(false);
+        private object _trackerLock = new object();
+        
+        public void Clear()
+        {
+            lock (_trackerLock)
+            {
+                _itemsInProcess.Clear();
+                _allItemsComplete.Reset();
+            }
+        }
+
+        public bool WaitForCompletion(int millisecondsTimeout)
+        {
+            return _allItemsComplete.WaitOne(millisecondsTimeout);
+        }
+
+        public void IssuePendingNotifications(ITestEventListener listener)
+        {
+            lock (_trackerLock)
+            {
+                int count = _itemsInProcess.Count;
+
+                // Signal completion of all pending suites, in reverse order
+                while (count > 0)
+                    listener.OnTestEvent(CreateNotification(_itemsInProcess[--count]));
+            }
+        }
+
+        private static string CreateNotification(XmlNode startElement)
+        {
+            bool isSuite = startElement.Name == "start-suite";
+
+            XmlNode notification = XmlHelper.CreateTopLevelElement(isSuite ? "test-suite" : "test-case");
+            if (isSuite)
+                notification.AddAttribute("type", startElement.GetAttribute("type"));
+            notification.AddAttribute("id", startElement.GetAttribute("id"));
+            notification.AddAttribute("name", startElement.GetAttribute("name"));
+            notification.AddAttribute("fullname", startElement.GetAttribute("fullname"));
+            notification.AddAttribute("result", "Failed");
+            notification.AddAttribute("label", "Cancelled");
+            XmlNode failure = notification.AddElement("failure");
+            XmlNode message = failure.AddElementWithCDataSection("message", "Test run cancelled by user");
+            return notification.OuterXml;
+        }
+
+        void ITestEventListener.OnTestEvent(string report)
+        {
+            XmlNode xmlNode = XmlHelper.CreateXmlNode(report);
+
+            lock (_trackerLock)
+            {
+                switch (xmlNode.Name)
+                {
+                    case "start-test":
+                    case "start-suite":
+                        _itemsInProcess.Add(xmlNode);
+                        break;
+
+                    case "test-case":
+                    case "test-suite":
+                        string id = xmlNode.GetAttribute("id");
+                        RemoveItem(id);
+
+                        if (_itemsInProcess.Count == 0)
+                            _allItemsComplete.Set();
```

Comment:
I'm not totally clear on how these ManualResetEvents work - but would this logic work for a test run where zero tests are actually run e.g. all filtered out?

Of course, you'd be hard-pressed to cancel it before it finished normally in that case! 😄

Reply:
Now that's niche! :smiley_cat: